syntax = "proto3";

import "constants.proto";
import "network_constants.proto";

message WifiScanStart {
    bool blocking = 1;
    bool passive = 2;
    uint32 group_channels = 3;
    uint32 period_ms = 4;
}

message ThreadScanStart {
    bool blocking = 1;
    uint32 channel_mask = 2;
}

message CmdScanStart {
    NetworkType net_type = 1;
    oneof payload {
        WifiScanStart wifi_scan_start = 10;
        ThreadScanStart thread_scan_start = 11;
    }
}

message RespScanStart {
    NetworkType net_type = 1;
}

message CmdScanStatus {
    NetworkType net_type = 1;
}

message RespScanStatus {
    NetworkType net_type = 1;
    bool scan_finished = 2;
    uint32 result_count = 3;
}

message CmdScanResult {
    NetworkType net_type = 1;
    uint32 start_index = 2;
    uint32 count = 3;
}

message WifiScanResult {
    bytes ssid = 1;
    uint32 channel = 2;
    int32 rssi = 3;
    bytes bssid = 4;
    WifiAuthMode auth = 5;
}

message ThreadScanResult {
    uint32 pan_id = 1;
    uint32 channel = 2;
    int32 rssi = 3;
    uint32 lqi = 4;
    bytes ext_addr = 5;
    string network_name = 6;
    bytes ext_pan_id = 7;
}

message ScanResult {
    oneof result {
        WifiScanResult  wifi_result= 10;
        ThreadScanResult thread_result = 11;
    }
}

message RespScanResult {
    NetworkType net_type = 1;
    repeated ScanResult entries = 2;
}

enum NetworkScanMsgType {
    TypeCmdScanStart = 0;
    TypeRespScanStart = 1;
    TypeCmdScanStatus = 2;
    TypeRespScanStatus = 3;
    TypeCmdScanResult = 4;
    TypeRespScanResult = 5;
}

message NetworkScanPayload {
    NetworkScanMsgType msg = 1;
    Status status = 2;
    oneof payload {
        CmdScanStart cmd_scan_start = 10;
        RespScanStart resp_scan_start = 11;
        CmdScanStatus cmd_scan_status = 12;
        RespScanStatus resp_scan_status = 13;
        CmdScanResult cmd_scan_result = 14;
        RespScanResult resp_scan_result = 15;
    }
}
